### Basic working of Google Palm LLM in LangChain

In [15]:
from langchain.llms import GooglePalm

api_key = 'AIzaSyDz2Dt2LonIglhqMihyIJiAFeIgT82p9XA'  # Get your API key from https://makersuite.google.com/

llm = GooglePalm(google_api_key=api_key, temperature=0.1)

In [52]:
import pprint
import google.generativeai as palm

palm.configure(api_key='AIzaSyDz2Dt2LonIglhqMihyIJiAFeIgT82p9XA')

In [3]:
models=[m for m in palm.list_models() if 'generateText' in m.supported_generation_methods]
model=models[0].name
print(model)

models/text-bison-001


In [7]:
import google.generativeai as genai

In [19]:



genai.configure(api_key="AIzaSyDz2Dt2LonIglhqMihyIJiAFeIgT82p9XA")

model = genai.GenerativeModel("gemini-pro")
response = model.generate_content("Write a 4-line poem of my love for samosa")
print(response.text)


Crispy shell, savory heart,
Samosa, my culinary art.
Golden brown, a symphony of spice,
A taste of heaven, a perfect slice.


In [10]:
essay = model.generate_content("write email requesting refund for electronic item")
print(essay)

response:
GenerateContentResponse(
    done=True,
    iterator=None,
    result=protos.GenerateContentResponse({
      "candidates": [
        {
          "content": {
            "parts": [
              {
                "text": "Dear [Customer Service Representative Name],\n\nI am writing to request a refund for an electronic item I purchased from your company on [date of purchase]. The order number is [order number].\n\nThe [item name] I received was defective upon arrival. I have attempted to troubleshoot the issue myself, but have been unable to resolve it. I have attached a copy of my troubleshooting log for your reference.\n\nI am very disappointed with this product, as I was really looking forward to using it. I would appreciate it if you could process a full refund for my purchase.\n\nI am available to [return the item or provide proof of the defect] at your convenience. Please let me know the best way to proceed.\n\nThank you for your attention to this matter.\n\nSincerely,\

In [44]:
from langchain.chains import RetrievalQA
from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### Now let's load data from Codebasics FAQ csv file

In [46]:
from langchain.document_loaders.csv_loader import CSVLoader
loader = CSVLoader(file_path="codebasics_faqs.csv", source_column="prompt", encoding="ISO-8859-1")
data = loader.load()
print(data)

[Document(metadata={'source': 'I have never done programming in my life. Can I take this bootcamp?', 'row': 0}, page_content='prompt: I have never done programming in my life. Can I take this bootcamp?\nresponse: Yes, this is the perfect bootcamp for anyone who has never done coding and wants to build a career in the IT/Data Analytics industry or just wants to perform better in your current job or business using data.'), Document(metadata={'source': 'Why should I trust Codebasics?', 'row': 1}, page_content='prompt: Why should I trust Codebasics?\nresponse: Till now 9000 + learners have benefitted from the quality of our courses. You can check the review section and also we have attached their LinkedIn profiles so that you can connect with them and ask directly.'), Document(metadata={'source': 'Is there any prerequisite for taking this bootcamp ?', 'row': 2}, page_content='prompt: Is there any prerequisite for taking this bootcamp ?\nresponse: Our bootcamp is specifically designed for b

### Hugging Face Embeddings

In [22]:
pip install torch transformers sentence-transformers langchain


   ---------------------------------------- 0.0/204.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/204.1 MB ? eta -:--:--
   ---------------------------------------- 0.5/204.1 MB 1.7 MB/s eta 0:02:02
   ---------------------------------------- 0.8/204.1 MB 1.5 MB/s eta 0:02:14
   ---------------------------------------- 1.0/204.1 MB 1.5 MB/s eta 0:02:18
   ---------------------------------------- 1.3/204.1 MB 1.5 MB/s eta 0:02:20
   ---------------------------------------- 1.6/204.1 MB 1.6 MB/s eta 0:02:11
   ---------------------------------------- 2.1/204.1 MB 1.6 MB/s eta 0:02:10
   ---------------------------------------- 2.4/204.1 MB 1.7 MB/s eta 0:02:02
    --------------------------------------- 2.9/204.1 MB 1.7 MB/s eta 0:01:57
    --------------------------------------- 3.7/204.1 MB 1.9 MB/s eta 0:01:48
    --------------------------------------- 4.2/204.1 MB 2.0 MB/s eta 0:01:41
    --------------------------------------- 4.7/204.1 MB 2.1 MB/s eta 0:01:37


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [23]:
pip install torch

Note: you may need to restart the kernel to use updated packages.


In [25]:
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118

^C
Note: you may need to restart the kernel to use updated packages.


In [27]:
pip install --upgrade langchain langchain-community

   ---------------------------------------- 0.0/1.0 MB ? eta -:--:--
   ---------------------------------------- 0.0/1.0 MB ? eta -:--:--
   ---------------------------------------- 0.0/1.0 MB ? eta -:--:--
   ---------- ----------------------------- 0.3/1.0 MB ? eta -:--:--
   ---------- ----------------------------- 0.3/1.0 MB ? eta -:--:--
   -------------------- ------------------- 0.5/1.0 MB 672.2 kB/s eta 0:00:01
   ------------------------------- -------- 0.8/1.0 MB 799.2 kB/s eta 0:00:01
   ---------------------------------------- 1.0/1.0 MB 825.5 kB/s eta 0:00:00
   ---------------------------------------- 0.0/2.5 MB ? eta -:--:--
   ---------------------------------------- 0.0/2.5 MB ? eta -:--:--
   ---- ----------------------------------- 0.3/2.5 MB ? eta -:--:--
   -------- ------------------------------- 0.5/2.5 MB 799.2 kB/s eta 0:00:03
   -------- ------------------------------- 0.5/2.5 MB 799.2 kB/s eta 0:00:03
   ------------ --------------------------- 0.8/2.5 MB 729

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [31]:
%pip install --upgrade --quiet  langchain-google-genai

Note: you may need to restart the kernel to use updated packages.


In [13]:
import getpass
import os

if "GOOGLE_API_KEY" not in os.environ:
    os.environ["GOOGLE_API_KEY"] = 'getpass.getpass("AIzaSyDz2Dt2LonIglhqMihyIJiAFeIgT82p9XA")'

In [39]:
from langchain_google_genai import GoogleGenerativeAIEmbeddings

embeddings = GoogleGenerativeAIEmbeddings(
    model="models/embedding-001",  # Corrected model name
    google_api_key="AIzaSyDz2Dt2LonIglhqMihyIJiAFeIgT82p9XA"  # Ensure correct key format
)

vector = embeddings.embed_query("hello, world!")
print(vector[:5])  # Print first 5 values

[0.05168594419956207, -0.030764883384108543, -0.03062233328819275, -0.02802734263241291, 0.01813093200325966]


In [40]:
vectors = embeddings.embed_documents(
    [
        "Today is Monday",
        "Today is Tuesday",
        "Today is April Fools day",
    ]
)
len(vectors), len(vectors[0])

(3, 768)

In [31]:
help(embeddings)

Help on GoogleGenerativeAIEmbeddings in module langchain_google_genai.embeddings object:

class GoogleGenerativeAIEmbeddings(pydantic.main.BaseModel, langchain_core.embeddings.embeddings.Embeddings)
 |  GoogleGenerativeAIEmbeddings(*, client: Any = None, model: str, task_type: Optional[str] = None, google_api_key: Optional[pydantic.types.SecretStr] = None, credentials: Any = None, client_options: Optional[Dict] = None, transport: Optional[str] = None, request_options: Optional[Dict] = None) -> None
 |
 |  `Google Generative AI Embeddings`.
 |
 |  To use, you must have either:
 |
 |      1. The ``GOOGLE_API_KEY`` environment variable set with your API key, or
 |      2. Pass your API key using the google_api_key kwarg
 |      to the GoogleGenerativeAIEmbeddings constructor.
 |
 |  Example:
 |      .. code-block:: python
 |
 |          from langchain_google_genai import GoogleGenerativeAIEmbeddings
 |
 |          embeddings = GoogleGenerativeAIEmbeddings(model="models/embedding-001")
 | 

In [18]:
model="models/text-embedding-001"

In [19]:
model="text-embedding-001"

In [26]:
pip install --upgrade langchain langchain-google-genai google-generativeai

Note: you may need to restart the kernel to use updated packages.


As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

### Vector store using FAISS

In [1]:
from langchain.document_loaders.csv_loader import CSVLoader
from langchain.vectorstores import FAISS
from langchain_google_genai import GoogleGenerativeAIEmbeddings


loader = CSVLoader(file_path="codebasics_faqs.csv", source_column="prompt", encoding="ISO-8859-1")
data = loader.load()
print(f"Loaded {len(data)} documents.")


embedding_model = GoogleGenerativeAIEmbeddings(
    model="models/embedding-001", 
    google_api_key="AIzaSyDz2Dt2LonIglhqMihyIJiAFeIgT82p9XA"
)

vectordb = FAISS.from_documents(documents=data, embedding=embedding_model)


retriever = vectordb.as_retriever(score_threshold=0.7)

query = "How do I use FAISS in LangChain?"
results = retriever.get_relevant_documents(query)

for doc in results:
    print(doc.page_content)


Loaded 75 documents.


C:\Users\deeks\AppData\Local\Temp\ipykernel_22960\1697016426.py:22: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  results = retriever.get_relevant_documents(query)


prompt: I am not allowing to post doubt in the discord group
response: Sure I can guide you

Go to the 'click-here-to-ask-questions' section and verify yourself here by clicking on the checkmark.
prompt: How can I get help if I have a doubt and need support?
response: We have an active discord server where you can post your question and most of the time you will get an answer in a reasonable time frame.
prompt: How can I contact the instructors for any doubts/support?
response: We have created every lecture with a motive to explain everything in an easy-to-understand manner. While working on these lectures you could make mistakes in steps or have some doubts. You need to commit yourself to hold patience, make efforts & diagnose the errors yourself by googling in order to become truly job ready. For any questions, that Google cannot answer or if you hit a wall - we got you covered! You can join our active discord community. which is a dedicated platform to discuss & clear your doubts wi

In [55]:
!pip install faiss-cpu

   ---------------------------------------- 0.0/13.8 MB ? eta -:--:--
   ---------------------------------------- 0.0/13.8 MB ? eta -:--:--
    --------------------------------------- 0.3/13.8 MB ? eta -:--:--
   - -------------------------------------- 0.5/13.8 MB 1.4 MB/s eta 0:00:10
   -- ------------------------------------- 0.8/13.8 MB 1.1 MB/s eta 0:00:12
   --- ------------------------------------ 1.0/13.8 MB 1.2 MB/s eta 0:00:11
   --- ------------------------------------ 1.3/13.8 MB 1.3 MB/s eta 0:00:10
   ---- ----------------------------------- 1.6/13.8 MB 1.3 MB/s eta 0:00:10
   ------ --------------------------------- 2.1/13.8 MB 1.4 MB/s eta 0:00:09
   ------ --------------------------------- 2.4/13.8 MB 1.5 MB/s eta 0:00:08
   -------- ------------------------------- 2.9/13.8 MB 1.5 MB/s eta 0:00:08
   --------- ------------------------------ 3.4/13.8 MB 1.7 MB/s eta 0:00:07
   ----------- ---------------------------- 3.9/13.8 MB 1.7 MB/s eta 0:00:06
   ------------ ----

In [58]:
pip install faiss-gpu

Note: you may need to restart the kernel to use updated packages.


ERROR: Could not find a version that satisfies the requirement faiss-gpu (from versions: none)
ERROR: No matching distribution found for faiss-gpu


In [59]:
pip install faiss-cpu


  Using cached faiss_cpu-1.9.0.post1-cp312-cp312-win_amd64.whl.metadata (4.5 kB)
Using cached faiss_cpu-1.9.0.post1-cp312-cp312-win_amd64.whl (13.8 MB)


In [2]:
rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

[Document(id='07be2ac6-ada2-447f-98f3-44210d68d728', metadata={'source': 'Do you provide any job assistance?', 'row': 11}, page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.'),
 Document(id='bc9ed30c-7460-4978-bff8-b4c6d10dcb0b', metadata={'source': 'Will this bootcamp guarantee me a job?', 'row': 15}, page_content='prompt: Will this bootcamp guarantee me a job?\nresponse: The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.'),
 Document(id='c0ce

As you can see above, the retriever that was created using FAISS and hugging face embedding is now capable of pulling relavant documents from our original CSV file knowledge store. This is very powerful and it will help us further in our project

##### Embeddings can be created using GooglePalm too. Also for vector database you can use chromadb as well as shown below. During our experimentation, we found hugging face embeddings and FAISS to be more appropriate for our use case

In [14]:
# google_palm_embeddings = GooglePalmEmbeddings(google_api_key=api_key)

# from langchain.vectorstores import Chroma
# vectordb = Chroma.from_documents(data,
#                            embedding=google_palm_embeddings,
#                            persist_directory='./chromadb')
# vectordb.persist()

### Create RetrievalQA chain along with prompt template üöÄ

In [6]:
from langchain.prompts import PromptTemplate
from langchain.chains import RetrievalQA

# ‚úÖ Define prompt template
prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer, try to provide as much text as possible from the "response" section in the source document without making many changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""

PROMPT = PromptTemplate(template=prompt_template, input_variables=["context", "question"])
chain_type_kwargs = {"prompt": PROMPT}

# ‚úÖ Initialize LLM (Choose one: Gemini or OpenAI GPT)
from langchain_google_genai import ChatGoogleGenerativeAI
llm = ChatGoogleGenerativeAI(model="gemini-pro", google_api_key="AIzaSyDz2Dt2LonIglhqMihyIJiAFeIgT82p9XA")

# ‚úÖ Ensure retriever is defined
retriever = vectordb.as_retriever()

# ‚úÖ Create the RetrievalQA chain
chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=retriever,
    return_source_documents=True,
    chain_type_kwargs=chain_type_kwargs
)

### We are all set 👍🏼 Let's ask some questions now

In [7]:
chain('Do you provide job assistance and also do you provide job gurantee?')

{'query': 'Do you provide job assistance and also do you provide job gurantee?',
 'result': "I don't know.",
 'source_documents': [Document(id='07be2ac6-ada2-447f-98f3-44210d68d728', metadata={'source': 'Do you provide any job assistance?', 'row': 11}, page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.'),
  Document(id='bc9ed30c-7460-4978-bff8-b4c6d10dcb0b', metadata={'source': 'Will this bootcamp guarantee me a job?', 'row': 15}, page_content='prompt: Will this bootcamp guarantee me a job?\nresponse: The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the j

**As you can see above, the answer of question comes from two different FAQs within our csv file and it is able to pull those questions and merge them nicely**

In [8]:
chain("Do you guys provide internship and also do you offer EMI payments?")

{'query': 'Do you guys provide internship and also do you offer EMI payments?',
 'result': "I don't know.",
 'source_documents': [Document(id='864d855d-9b99-433b-8c46-64b110cff07f', metadata={'source': 'Do we have an EMI option?', 'row': 13}, page_content='prompt: Do we have an EMI option?\nresponse: No'),
  Document(id='b1f5e646-4af6-4320-bb9d-dd90f1259c9c', metadata={'source': 'Do you provide any virtual internship?', 'row': 14}, page_content='prompt: Do you provide any virtual internship?\nresponse: Yes'),
  Document(id='07be2ac6-ada2-447f-98f3-44210d68d728', metadata={'source': 'Do you provide any job assistance?', 'row': 11}, page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.'),
  Document(id='4b391727-e3bb-4c86-ba85-6c12524e3b60', metadata={'source': 'Can I add this course to my r

In [9]:
chain("do you have javascript course?")

{'query': 'do you have javascript course?',
 'result': "I don't know. The provided context does not mention anything about JavaScript courses.",
 'source_documents': [Document(id='919df603-6243-41df-a54f-229037f08fc3', metadata={'source': '\nI don\x92t have a laptop, can I take this course?', 'row': 25}, page_content='prompt: I don\x92t have a laptop, can I take this course?\nresponse: We recommend learning by doing and therefore you need to have a laptop or a PC (at least 4 GB ram).'),
  Document(id='a5eb5228-3c39-4252-a19f-5e7f00245411', metadata={'source': 'Can I add this course to my resume? If Yes, how?', 'row': 34}, page_content='prompt: Can I add this course to my resume? If Yes, how?\nresponse: Absolutely, we have a section in this course explaining how you can add the learnings from this course in your resume that will appeal to the hiring team.'),
  Document(id='b6a6226a-9198-410b-9ad0-9d37f1d30a03', metadata={'source': 'Why should I trust Codebasics?', 'row': 1}, page_conten

In [10]:
chain("Do you have plans to launch blockchain course in future?")

{'query': 'Do you have plans to launch blockchain course in future?',
 'result': "I don't know.",
 'source_documents': [Document(id='a5eb5228-3c39-4252-a19f-5e7f00245411', metadata={'source': 'Can I add this course to my resume? If Yes, how?', 'row': 34}, page_content='prompt: Can I add this course to my resume? If Yes, how?\nresponse: Absolutely, we have a section in this course explaining how you can add the learnings from this course in your resume that will appeal to the hiring team.'),
  Document(id='4b391727-e3bb-4c86-ba85-6c12524e3b60', metadata={'source': 'Can I add this course to my resume?', 'row': 19}, page_content='prompt: Can I add this course to my resume?\nresponse: Yes. Absolutely you can mention the AtliQ Hardware project experience in your resume with the relevant skills that you will learn from this course'),
  Document(id='b1f5e646-4af6-4320-bb9d-dd90f1259c9c', metadata={'source': 'Do you provide any virtual internship?', 'row': 14}, page_content='prompt: Do you pro

In [11]:
chain("should I learn power bi or tableau?")

{'query': 'should I learn power bi or tableau?',
 'result': "I don't know.",
 'source_documents': [Document(id='35dab637-5818-4c9b-a550-fd7d99cd1cb5', metadata={'source': '\nPower BI or Tableau which one is better?', 'row': 29}, page_content='prompt: Power BI or Tableau which one is better?\nresponse: This is a contextual question. If you are talking about a pure visualization tool Tableau is slightly better. Data connectors, modeling and transformation features are available in both. However, factually speaking Power BI is cheaper and offers tighter integration with the Microsoft environment. Since most companies use excel & Microsoft tools they start with Power BI or move towards Power BI for seamless integration with other Microsoft tools (called as Power platform). This makes the job openings grow at a much higher rate on Power BI and Power Platform. Also, Power BI has been leading the Gartner\x92s magic quadrant in BI for the last few years as the industry leader.'),
  Document(id

In [12]:
chain("I've a MAC computer. Can I use powerbi on it?")

{'query': "I've a MAC computer. Can I use powerbi on it?",
 'result': 'Power BI desktop works only in Windows OS. However, you can use a virtual machine to install and work with Power BI in other Operating systems.',
 'source_documents': [Document(id='c6890116-f235-4a9d-b75e-470c408095e4', metadata={'source': 'How can I use PowerBI on my Mac system?', 'row': 44}, page_content='prompt: How can I use PowerBI on my Mac system?\nresponse: Hi\n\nYou can use VirtualBox to create a virtual machine and install Windows on it. This will allow you to run Power BI and Excel on your Mac.\n\nIf you\'re not familiar with setting up a virtual machine, there are many resources available on YouTube that can guide you through the process. Simply search for "installing virtual machines" and you\'ll find plenty of helpful videos.\n\nBest of luck with your studies!'),
  Document(id='ef424ddc-920b-42dd-b6ac-34fd0bb966bf', metadata={'source': 'Does Power BI work in Mac OS/Ubuntu?', 'row': 31}, page_content='p

In [13]:
chain("I don't see power pivot. how can I enable it?")

{'query': "I don't see power pivot. how can I enable it?",
 'result': "I don't know.",
 'source_documents': [Document(id='24c377f7-b8be-4e20-adfc-b36333aada80', metadata={'source': 'How do I enable Power Pivot before using it for the first time ?', 'row': 74}, page_content='prompt: How do I enable Power Pivot before using it for the first time ?\nresponse: Follow the process in the link : \n\nhttps://drive.google.com/file/d/1-mO-v52h-YTY1s-v30liBJPu6Yj4OUxb/view?usp=share_link'),
  Document(id='13b42a91-cd8f-4c54-930a-401431936bad', metadata={'source': 'How to install power pivot if its not available in system?', 'row': 38}, page_content='prompt: How to install power pivot if its not available in system?\nresponse: Follow this thread for instructions - https://support.microsoft.com/en-us/office/start-the-power-pivot-add-in-for-excel-a891a66d-36e3-43fc-81e8-fc4798f39ea8\nIf it doesn\'t show in the ribbon then go to "insert" tab. You will be able to see pivot table option there.'),
  Doc

In [14]:
chain("What is the price of your machine learning course?")

{'query': 'What is the price of your machine learning course?',
 'result': "I don't know.",
 'source_documents': [Document(id='51d3a91f-f68b-4bb3-bea2-62864e61fa67', metadata={'source': 'I\x92m not sure if this course is good enough for me to invest some money. What can I do?', 'row': 20}, page_content='prompt: I\x92m not sure if this course is good enough for me to invest some money. What can I do?\nresponse: Don\x92t worry. Many videos in this course are free so watch them to get an idea of the quality of teaching. Dhaval Patel (the course instructor) runs a popular data science YouTube channel called Codebasics. On that, you can watch his videos and read comments to get an idea of his teaching style'),
  Document(id='0a6d7c9d-08b8-46cf-a0e8-29433892a7e2', metadata={'source': 'Is there any prerequisite for taking this course?', 'row': 35}, page_content='prompt: Is there any prerequisite for taking this course?\nresponse: The only prerequisite is that you need to have a functional lap